bundle/config/mutator/run_as.go

```diff
@@ -23,12 +27,120 @@ func (m *setRunAs) Name() string {
+//     deployment user. For example, jobs.
+//  2. Does not make sense for these resources to run_as a different user. We do not
+//     have plans to add platform side support for `run_as` for these resources.
+//     For example, experiments or registered models.
+var allowListForRunAsOther = []string{"jobs", "models", "registered_models", "experiments"}
+
+// Resources that do not allow setting a run_as identity to a different user but
+// have plans to add platform side support for `run_as` for these resources at
+// some point in the future. For example, pipelines, model serving endpoints or lakeview dashboards.
+//
+// We expect the allow list and the deny list to form mutually exclusive and exhaustive
+// sets of all resource types that are supported by DABs.
+var denyListForRunAsOther = []string{"pipelines", "model_serving_endpoints"}
+
+type errorUnsupportedResourceTypeForRunAs struct {
+	resourceType  string
+	resourceValue dyn.Value
+	currentUser   string
+	runAsUser     string
+}
+
+// TODO(6 March 2024): This error message is big. We should split this once
+// diag.Diagnostics is ready.
+// TODO(6 March 2024): Link the docs page describing run_as semantics in the error below
+// once the page is ready.
+func (e errorUnsupportedResourceTypeForRunAs) Error() string {
+	return fmt.Sprintf("%s are not supported when the current deployment user is different from the bundle's run_as identity. Please deploy as the run_as identity. List of supported resources: [%s]. Location of the unsupported resource: %s. Current identity: %s. Run as identity: %s", e.resourceType, strings.Join(allowListForRunAsOther, ", "), e.resourceValue.Location(), e.currentUser, e.runAsUser)
+}
+
+func getRunAsIdentity(runAs dyn.Value) (string, error) {
+	// Get service principal name and user name from run_as section
+	runAsSp, err := dyn.Get(runAs, "service_principal_name")
+	if err != nil && !dyn.IsNoSuchKeyError(err) {
+		return "", err
+	}
+	runAsUser, err := dyn.Get(runAs, "user_name")
+	if err != nil && !dyn.IsNoSuchKeyError(err) {
+		return "", err
+	}
+
+	sp, spIsDefined := runAsSp.AsString()
+	user, userIsDefined := runAsUser.AsString()
+
+	switch {
+	case spIsDefined && userIsDefined:
+		return "", fmt.Errorf("run_as section must specify exactly one identity. A service_principal_name %q is specified at %s. A user_name %q is defined at %s", sp, runAsSp.Location(), user, runAsUser.Location())
+	case spIsDefined:
+		return sp, nil
+	case userIsDefined:
+		return user, nil
+	default:
+		return "", nil
+	}
+}
+
 func (m *setRunAs) Apply(_ context.Context, b *bundle.Bundle) error {
+	// Return early if run_as is not defined in the bundle
 	runAs := b.Config.RunAs
 	if runAs == nil {
 		return nil
 	}
 
+	currentUser := b.Config.Workspace.CurrentUser.UserName
+
+	// Assert the run_as configuration is valid in the context of the bundle
+	err := b.Config.Mutate(func(v dyn.Value) (dyn.Value, error) {
```

Comment:
Can we just simplify it to ?
```
if b.Config.Resources.Pipelines != nil || b.Config.Resources.ModelServingEndpoints != nil {
  return errorUnsupportedResourceTypeForRunAs{...}
}
```

All the dynamic logic does not really help anyway because allow/deny list is static, it's just makes things more difficult to follow

Reply:
The main benefit is introspection into the `resources` block. The dynamic approach allows us to ensure new resources cannot be added to DABs without explicitly being added to either the allow list or deny list. 

If we use the strongly typed approach, we run the risk of allow listing a resource that does not support `run_as` yet but will at some point. Retroactively adding `run_as` support is not an option because it'll be a breaking change. 
A good example is lakeview dashboards, when they are added we need to have them in the deny list.

Reply:
This is codified in the `TestAllResourceTypesAreClassifiedForRunAs` test.

Reply:
> If we use the strongly typed approach, we run the risk of allow listing a resource that does not support

Not sure I agree with that. Applying top-level run_as permissions is done implicitly below, see `// Set run_as for jobs` section. If we add a new resource like Lakeview in DABs support, run_as won't be automatically applied

Reply:
So adding something to `denyListForRunAs ` is equal to adding `if b.Config.Resources.{resourcetype} != nil {return err}`

Reply:
Done, now we use typed if statements to do the validation, as proposed:
```
if b.Config.Resources.Pipelines != nil || b.Config.Resources.ModelServingEndpoints != nil {
  return errorUnsupportedResourceTypeForRunAs{...}
}
```